crates/sui-source-validation/src/lib.rs

```diff
@@ -396,3 +474,273 @@ fn local_modules(
+                        decode_bytecode_file(build_path.clone(), package, &bytecode_path)?;
+                    units.push((*package, local_unit))
+                }
+                continue;
+            }
+            Some(_) => {
+                // If a toolchain version for a previous compiler exists, we
+                // already added all compiled units to deps_compiled_units.
+                continue;
+            }
+            None => bail!("Expected a version for {package} but none found"),
+        }
+    }
+    Ok(units)
+}
+
+fn download_and_compile(
+    root: PathBuf,
+    ToolchainVersion {
+        compiler_version,
+        edition,
+        flavor,
+    }: &ToolchainVersion,
+    dep_name: &Symbol,
+) -> anyhow::Result<()> {
+    let binaries_path = &*MOVE_HOME; // E.g., ~/.move/binaries
+    let mut dest_dir = PathBuf::from(binaries_path);
+    dest_dir = dest_dir.join("binaries");
+    let dest_version = dest_dir.join(compiler_version.clone());
+    let platform = detect_platform()?;
+    let dest_binary = dest_version.join(format!("target/release/sui-{}", platform));
+    let dest_binary_os = OsStr::new(dest_binary.as_path());
+
+    if !dest_binary.exists() {
+        // Download if binary does not exist.
+        let url = format!("https://github.com/MystenLabs/sui/releases/download/mainnet-v{}/sui-mainnet-v{}-{}.tgz", compiler_version, compiler_version, platform);
+
+        println!(
+            "{} compiler @ {} (this may take a while)",
+            "DOWNLOADING".bold().green(),
+            compiler_version.yellow()
+        );
+
+        let mut response = ureq::get(&url).call()?.into_reader();
+        let dest_tarball = dest_version.join(format!("{}.tgz", compiler_version));
+        debug!("tarball destination: {} ", dest_tarball.display());
+        if let Some(parent) = dest_tarball.parent() {
+            std::fs::create_dir_all(parent)
+                .map_err(|e| anyhow!("failed to create directory for tarball: {e}"))?;
+        }
+        let mut dest_file = File::create(&dest_tarball)?;
+        io::copy(&mut response, &mut dest_file)?;
+
+        // Extract the tarball using the tar crate
+        let tar_gz = File::open(&dest_tarball)?;
+        let tar = flate2::read::GzDecoder::new(tar_gz);
+        let mut archive = Archive::new(tar);
+        archive
+            .unpack(&dest_version)
+            .map_err(|e| anyhow!("failed to untar compiler binary: {e}"))?;
+
+        set_executable_permission(dest_binary_os)?;
+    }
+
+    debug!(
+        "sui move build --default-move-edition {} --default-move-flavor {} -p {}",
+        edition.to_string().as_str(),
+        flavor.to_string().as_str(),
+        root.display()
+    );
+    info!(
+        "{} {} (compiler @ {})",
+        "BUILDING".bold().green(),
+        dep_name.as_str(),
+        compiler_version.yellow()
+    );
+    Command::new(dest_binary_os)
+        .args([
+            OsStr::new("move"),
```

Comment:
One thing to note is that this is going to build all of the dependencies package graph using the specified toolchain version, rather than just the package itself.  Is it possible to use the specified toolchain version to compile only the dependency in question, to avoid some accidentally quadratic behaviour here? cc @tnowacki / @cgswords in case you know of any special compiler incantations that would make this work.

This may also be an acceptable trade-off to begin with?